lib/src/main/kotlin/tree/AVLTree.kt

```diff
@@ -2,16 +2,142 @@ package tree
+            node.height = 1 + maxOf(height(node.left), height(node.right))
+        }
+    }
+
+    private fun getBalance(node: AVLTreeNode<K, V>): Int {
+        return height(node.left) - height(node.right)
+    }
+
+    private fun rotateRight(node: AVLTreeNode<K, V>): AVLTreeNode<K, V>? {
+        val nodeLeft = node.left
+        val nodeLeftRight = nodeLeft?.right
+
+        nodeLeft?.right = node
+        node.left = nodeLeftRight
+
+        updateHeight(node)
+        updateHeight(nodeLeft)
+
+        return nodeLeft
+    }
+
+    private fun rotateLeft(node: AVLTreeNode<K, V>): AVLTreeNode<K, V>? {
+        val nodeRight = node.right
+        val nodeRightLeft = nodeRight?.left
+
+        nodeRight?.left = node
+        node.right = nodeRightLeft
+
+        updateHeight(node)
+        updateHeight(nodeRight)
+
+        return nodeRight
+    }
+
+    private fun rebalanced(root: AVLTreeNode<K, V>?, node: AVLTreeNode<K, V>?): AVLTreeNode<K, V>?{
+        updateHeight(root)
+        val balance = root?.let { getBalance(it) }
+        val rootLeft = root?.left
+        val rootRight = root?.right
+
+        if (node != null && balance != null){
+            if (balance > 1) {
+                if (rootLeft != null) {
+                    if (node.key < rootLeft.key) {
+                        return rotateRight(root)
+                    } else {
+                        root.left = root.left?.let { rotateLeft(it) }
+                        return rotateRight(root)
+                    }
+                }
+            }
+            if (balance < -1) {
+                if (rootRight != null) {
+                    if (node.key > rootRight.key) {
+                        return rotateLeft(root)
+                    } else {
+                        root.right = root.right?.let { rotateRight(it) }
+                        return rotateLeft(root)
+                    }
+                }
+            }
+        }
+
+        return root
+    }
+
+    private fun insertRecursive(root: AVLTreeNode<K, V>?, node: AVLTreeNode<K, V>?): AVLTreeNode<K, V>? {
+        if (root == null) {
+            return node
+        }
+
+        if (node != null) {
+            if (root.key > node.key) {
+                root.left = insertRecursive(root.left, node)
+            } else  if (root.key < node.key) {
+                    root.right = insertRecursive(root.right, node)
+            }
+        }
+
+        return rebalanced(root, node)
+    }
+
     override fun insertNode(node: AVLTreeNode<K, V>) {
-        TODO("Insert node to tree")
+        root = insertRecursive(root, node)
+    }
+
+    private fun minValueNode(node: AVLTreeNode<K, V>): AVLTreeNode<K, V> {
+        var current = node
+        val curLeft = current.left
+        while (curLeft != null) current = curLeft
+        return current
+    }
+    private fun removeRecursive(root: AVLTreeNode<K, V>?, node: AVLTreeNode<K, V>): AVLTreeNode<K, V>? {
+        var rootNode = root
```

Comment:
Методы лучше отделить новой строкой

```suggestion
    private fun minValueNode(node: AVLTreeNode<K, V>): AVLTreeNode<K, V> {
        var current = node
        val curLeft = current.left
        while (curLeft != null) current = curLeft
        return current
    }
    
    private fun removeRecursive(root: AVLTreeNode<K, V>?, node: AVLTreeNode<K, V>): AVLTreeNode<K, V>? {
        var rootNode = root
```